### DISCUSSION TITLE
NAJ With JS Contacts vs. Rust and AS

### DISCUSSION THREAD

**Discussion by BenKurrek** [near/near-sdk-js#28] [General]
Have you thought about the differences between integration using JS vs. Rust and AS? I think one of the beautiful things about NEAR is that once I deploy a smart contract, it doesn’t matter if it’s assembly script or rust. I could use the same frontend code whether it’s rust or assembly script. I think one of the painpoints for devs will be that given an arbitrary frontend (especially if they’re using our examples), they can’t just plug in any contract now. They would have to change the way the frontend is coded depending on if it’s rust / AS or JS due to the enclave approach.

**Comment by austinbaggio** [comment]
@itegulov have we thought about how an ABI might work with JS contracts?

**Reply by itegulov** [reply]
We have very briefly touched on this during offsite, but nothing has been decided yet. Basically given that JSVM contracts are being interacted with in a completely different way, there would have to be a special clause in the schema format for them (or even a completely different schema format).

CC @austinabell

**Reply by austinabell** [reply]
The takeaway was basically that the schema format would have a field that can be used ambiguously and the JS SDK can define its own semantics or we could have a standardization for these enclaves. 

This will likely not happen with the initial version of the ABI since schema generation from the JS SDK seems a bit far away, but is something that is planned to be taken into consideration.

**Comment by volovyks** [comment]
@BenKurrek I agree with you 100%. If we will stick to the enslaved design - we need to add more APIs to NAJ. Something similar to the new `NEAR CLI` commands.
Meanwhile, we need to have tutorials/examples on how you can call enslaved contracts in your dApp. Do we have such an example, or are we planning to create one? I think it is critical for the upcoming conference and hackathon.

**Reply by BenKurrek** [reply]
There's [this issue](https://github.com/near/near-api-js/issues/867) I've created in near-api-js, and I've also created [this game](https://github.com/BenKurrek/near-sdk-js-game) which includes a frontend. I just need to finish the documentation for it.

**Reply by volovyks** [reply]
Awesome, that issue describes a lot on how you need to use NAJ now. We just need to make that info more accessible.
